psi4/src/psi4/libfock/DirectJK.cc

```diff
@@ -720,88 +768,762 @@ void DirectJK::build_JK(std::vector<std::shared_ptr<TwoBodyAOInt>>& ints, std::v
+            size += primary_->shell(P).nfunction();
+        }
+        max_basis_buffer = std::max(max_basis_buffer, size);
+    }
+
+    if (debug_) {
+        outfile->Printf("  ==> LinK: Atom Blocking <==\n\n");
+        for (size_t atom = 0; atom < natom; atom++) {
+            outfile->Printf("  Atom: %3d, Atom Start: %4d, Atom End: %4d\n", atom, atom_first_shell_idx[atom],
+                            atom_first_shell_idx[atom + 1]);
+            for (int P = atom_first_shell_idx[atom]; P < atom_first_shell_idx[atom + 1]; P++) {
+                int size = primary_->shell(P).nfunction();
+                int off = primary_->shell(P).function_index();
+                int off2 = shell_first_basis_idx[P];
+                outfile->Printf("    Shell: %4d, Size: %4d, Offset: %4d, Offset2: %4d\n", P, size, off,
+                                off2);
+            }
+        }
+        outfile->Printf("\n");
+    }
+
+    // => Significant Atom Pairs (PQ|-style <= //
+
+    std::vector<std::pair<int, int>> atom_pairs;
+    for (size_t Patom = 0; Patom < natom; Patom++) {
+        for (size_t Qatom = 0; Qatom < natom; Qatom++) {
+            if (Qatom > Patom) continue;
+            bool found = false;
+            for (int P = atom_first_shell_idx[Patom]; P < atom_first_shell_idx[Patom + 1]; P++) {
+                for (int Q = atom_first_shell_idx[Qatom]; Q < atom_first_shell_idx[Qatom + 1]; Q++) {
+                    if (ints[0]->shell_pair_significant(P, Q)) {
+                        found = true;
+                        atom_pairs.push_back(std::pair<int, int>(Patom, Qatom));
+                        break;
+                    }
+                }
+                if (found) break;
+            }
+        }
+    }
+
+    // => Calculate Shell Ceilings (For significant bra-ket atom pairs)
+    std::vector<double> shell_ceilings(nshell);
+    for (int P = 0; P < nshell; P++) {
+        for (int Q = 0; Q < nshell; Q++) {
+            double val = std::sqrt(ints[0]->shell_ceiling2(P, Q, P, Q));
+            shell_ceilings[P] = std::max(shell_ceilings[P], val);
+        }
+    }
+
+    // => Signficant Ket Atoms for a given Bra Atom <= //
+    std::vector<std::unordered_set<int>> significant_kets(natom);
+    for (size_t Patom = 0; Patom < natom; Patom++) {
+        for (size_t Ratom = 0; Ratom < natom; Ratom++) {
+            bool found = false;
+            for (int P = atom_first_shell_idx[Patom]; P < atom_first_shell_idx[Patom + 1]; P++) {
+                for (int R = atom_first_shell_idx[Ratom]; R < atom_first_shell_idx[Ratom + 1]; R++) {
+                    double screen = shell_ceilings[P] * shell_ceilings[R] * ints[0]->shell_pair_max_density(P, R);
+                    if (screen >= linK_ints_cutoff_) {
+                        found = true;
+                        significant_kets[Patom].emplace(Ratom);
+                        break;
+                    }
+                }
+                if (found) break;
```

Comment:
Ah, yes, you pointed out a bug, I would need to run the mini-list code with PR, PS, QR, QS (4 loops), and not 2.
Currently, I just track PR and QR. See figure 1 in the Oschenfeld paper, as well as the discussion surrounding mini-lists below the figure 

"note that in the present implementation only one final ML and an N^2 K-storage is used. Although this is not important even for very large systems, it can be easily avoided. To ensure the formation directly into a sparse K, four minilists are required, which is only a trivial extension of the algorithm depicted in Fig. 1"